### PR TITLE
[MRG] Allow write_raw_bids to specify a file format explicitly

### PR DESCRIPTION
**Pull request by adam2392**
PR Description
--------------

Closes: #659 

Note: EDF format is not supported due to the lack of a robust EDF format writer.

Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [ ] All comments resolved
- [ ] This is not your own PR
- [ ] All CIs are happy
- [ ] PR title starts with [MRG]
- [ ] [whats_new.rst](https://github.com/mne-tools/mne-bids/blob/master/doc/whats_new.rst) is updated
- [ ] PR description includes phrase "closes <#issue-number>"

### CODE REVIEW

**Review by agramfort — APPROVED** [review]
LGTM

**Comment by adam2392** [comment]
Not sure what's up w/ the CI but everything works locally and increased the test coverage.

File: mne_bids/write.py [hunk 1]
```diff
@@ -1095,6 +1104,11 @@ def write_raw_bids(raw, bids_path, events_data=None,
         raise RuntimeError('You passed event_id, but no events_data NumPy '
                            'array. You need to pass both, or neither.')
 
+    if format not in CONVERT_FORMATS:
+        raise ValueError(f'The "format" {format} is not an '
+                         f'accepted format. Please use one of '
+                         f'{CONVERT_FORMATS}')
```
Comment (sappelhoff) [inline mne_bids/write.py:1110]: This check should come later after line 1286 IMHO ... for example if a user specifies `BrainVision` but has MEG data, we should raise an error as well.

Finally, the current warning is confusing: you are talking about "accepted format", which will remind users of accepted formats in BIDS ... but what **you** actually mean is accepted input parameters to the write_raw_bids function.

Imagine I pass "EDF" ... you tell me it's not an accepted format ... but I know that it is. --> the warning msg should be something like "not supported input, please use ... "

**Review by sappelhoff — CHANGES_REQUESTED** [review]
LGTM in principle. I suggest we wait for #670, one other suggestion, and one requested change.

File: mne_bids/write.py [hunk 2]
```diff
@@ -1254,6 +1263,21 @@ def write_raw_bids(raw, bids_path, events_data=None,
...
+            bids_path.update(extension='.vhdr')
+        elif format == 'FIF' and bids_path.datatype == 'meg':
+            convert = True
+            bids_path.update(extension='.fif')
+        elif format not in CONVERT_FORMATS:
+            raise ValueError(f'The input "format" {format} is not an '
+                             f'accepted input format for `write_raw_bids`. '
+                             f'For example, if you passed in "EDF", mne-bids '
+                             f'does not currently have conversion to EDF. '
+                             f'Please use one of {CONVERT_FORMATS}.')
```
Comment (adam2392) [inline mne_bids/write.py:1279]: @sappelhoff

File: mne_bids/write.py (hunk 1 above)
Comment (adam2392) [reply mne_bids/write.py]: This is fixed in the latest commit. Lmk wyt.

**Review by agramfort — APPROVED** [review]
@sappelhoff floor is yours

File: mne_bids/write.py (hunk 2 above)
Comment (sappelhoff) [reply mne_bids/write.py:1279]: Looks better, but you are not catching the case where format is Brainvision, and Datatype is meg .... Or format is for and Datatype is eeg.

The rest looks good

**Review by sappelhoff — COMMENTED** [review]
One more comment

File: mne_bids/write.py (hunk 2 above)
Comment (adam2392) [reply mne_bids/write.py]: added in new commit.

File: mne_bids/write.py [hunk 3]
```diff
@@ -1276,6 +1271,12 @@ def write_raw_bids(raw, bids_path, events_data=None,
         elif format == 'FIF' and bids_path.datatype == 'meg':
             convert = True
             bids_path.update(extension='.fif')
+        elif format not in CONVERT_FORMATS:
+            raise ValueError(f'The input "format" {format} is not an '
+                             f'accepted input format for `write_raw_bids`. '
+                             f'For example, if you passed in "EDF", mne-bids '
+                             f'does not currently have conversion to EDF. '
```
Comment (hoechenberger) [inline mne_bids/write.py:1278]: I don't think this example is necessary, or is it? I assume it would only cause confusion

**Review by hoechenberger — APPROVED** [review]

**Review by sappelhoff — APPROVED** [review]
thanks a lot!

Just one conflict, then good to go!

**Comment by hoechenberger** [comment]
I've resolved the conflict, you can merge once CI goes green

**Comment by agramfort** [comment]
@adam2392 https://github.com/mne-tools/mne-bids/pull/672/checks?check_run_id=1655522660 failed

**Comment by hoechenberger** [comment]
I've restarted CI, the error @agramfort noticed is sometimes caused by flaky servers when we're downloading the test data. Let's see…

**Comment by sappelhoff** [comment]
>  flaky servers when we're downloading the test data.

I see an increasing frequency of this happening ... better to keep an eye on it.

**Pull request merged**